core/src/either.rs

```diff
@@ -57,41 +57,24 @@ where
 
 /// Implements `AsyncRead` and `AsyncWrite` and dispatches all method calls to
 /// either `First` or `Second`.
+#[pin_project]
 #[derive(Debug, Copy, Clone)]
 pub enum EitherOutput<A, B> {
-    First(A),
-    Second(B),
+    First(#[pin] A),
+    Second(#[pin] B),
 }
 
 impl<A, B> AsyncRead for EitherOutput<A, B>
 where
     A: AsyncRead,
     B: AsyncRead,
 {
-    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
-        match self {
-            EitherOutput::First(a) => a.prepare_uninitialized_buffer(buf),
-            EitherOutput::Second(b) => b.prepare_uninitialized_buffer(buf),
-        }
-    }
-
-    fn read_buf<Bu: bytes::BufMut>(&mut self, buf: &mut Bu) -> Poll<usize, IoError> {
-        match self {
-            EitherOutput::First(a) => a.read_buf(buf),
-            EitherOutput::Second(b) => b.read_buf(buf),
-        }
-    }
-}
-
-impl<A, B> Read for EitherOutput<A, B>
-where
-    A: Read,
-    B: Read,
-{
-    fn read(&mut self, buf: &mut [u8]) -> Result<usize, IoError> {
-        match self {
-            EitherOutput::First(a) => a.read(buf),
-            EitherOutput::Second(b) => b.read(buf),
+    #[project]
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, IoError>> {
```

Comment:
We should also override [`AsyncRead::poll_read_vectored`](https://docs.rs/futures/0.3.1/futures/prelude/trait.AsyncRead.html#method.poll_read_vectored).